### PR TITLE
feat: add per-model FP8 layerwise casting for VRAM reduction

### PR DESCRIPTION
**Pull request by Pfannkuchensack**
# FP8 Layerwise Casting - Implementation

## Summary

Add per-model `fp8_storage` option to model default settings that enables diffusers' `enable_layerwise_casting()` to store weights in FP8 (`float8_e4m3fn`) while casting to fp16/bf16 during inference. This reduces VRAM usage by ~50% per model with minimal quality loss.

Supported: SD1/SD2/SDXL/SD3, Flux, Flux2, CogView4, Z-Image, ControlNet, T2IAdapter.
Not applicable: Text Encoders, LoRA, GGUF, BnB, custom classes.

## Related Issues / Discussions

- https://github.com/invoke-ai/InvokeAI/issues/7148
- Based on approach from https://github.com/AUTOMATIC1111/stable-diffusion-webui/pull/14031
- Uses diffusers' native `enable_layerwise_casting()` (available in diffusers 0.36.0)

## QA Instructions

1. Set `fp8_storage: true` in a model's `default_settings` (via API or Model Manager UI)
2. Load the model and generate an image
3. Verify VRAM usage is reduced compared to normal loading
4. Verify image quality is acceptable (minimal degradation expected)
5. Verify Text Encoders are NOT affected (excluded by submodel type filter)
6. Verify non-CUDA devices gracefully ignore the setting

### Test Matrix

- [x] SD1.5 Diffusers with `fp8_storage=true` - load and generate
- [x] SDXL Diffusers with `fp8_storage=true` - load and generate
- [x] SDXL Safetensors with `fp8_storage=true` - load and generate
- [x] Flux1 Diffusers with `fp8_storage=true` - load and generate
- [x] Flux1 Safetensors with `fp8_storage=true` - load and generate
- [x] Flux2klein Diffusers with `fp8_storage=true` - load and generate
- [x] Flux2klein Safetensors with `fp8_storage=true` - load and generate
- [x] CogView4 with `fp8_storage=true` - load and generate
- [x] ControlNet with `fp8_storage=true` - load and generate
- [x] VRAM comparison: with vs. without `fp8_storage`
- [x] Image quality comparison: FP8 vs fp16/bf16
- [ ] MPS/CPU: verify `fp8_storage` is silently ignored
- [x] Text Encoder submodels: verify FP8 is NOT applied
- [x] GGUF/BnB models: verify FP8 is gracefully skipped
- [ ] Z-Image Turbo Diffusers with `fp8_storage=true` - load and generate # it does not work

## Checklist

- [x] _The PR has a short but descriptive title, suitable for a changelog_
- [x] _Tests added / updated (if applicable)_
- [ ] _Changes to a redux slice have a corresponding migration_
- [x] _Documentation added / updated (if applicable)_
- [ ] _Updated `What's New` copy (if doing a release after this PR)_

### CODE REVIEW

File: invokeai/backend/model_manager/load/load_default.py [hunk 1]
```diff
@@ -124,6 +124,54 @@ def get_size_fs(
             variant=config.repo_variant if isinstance(config, Diffusers_Config_Base) else None,
         )
 
+    def _should_use_fp8(self, config: AnyModelConfig, submodel_type: Optional[SubModelType] = None) -> bool:
+        """Check if FP8 layerwise casting should be applied to a model."""
+        # FP8 storage only works on CUDA
+        if self._torch_device.type != "cuda":
```
Comment (JPPhoto) [inline invokeai/backend/model_manager/load/load_default.py:160]: Shouldn't this check `self._get_execution_device()` to make sure the model is to be executed on cuda?

File: invokeai/backend/model_manager/load/model_loaders/flux.py [hunk 2]
```diff
@@ -139,6 +139,7 @@ def _load_model(
             local_files_only=True,
         )
 
+        model = self._apply_fp8_layerwise_casting(model, config, submodel_type)
```
Comment (JPPhoto) [inline invokeai/backend/model_manager/load/model_loaders/flux.py:142]: Does this only apply to the v2 VAE and diffusers models? What about GGUF?

**Review by JPPhoto — CHANGES_REQUESTED** [review]
In my quantized Krea dev setup, your code was never called - is this by design or an overlooked class?

I'd also like the UI to be tweaked so the fp8 setting appears as a single slider under Settings like CPU-only for text encoders rather than as a dual-slider in the model defaults section.

File: invokeai/backend/model_manager/load/load_default.py (hunk 1 above)
Comment (Pfannkuchensack) [reply invokeai/backend/model_manager/load/load_default.py:160]: It checks "does the system even have CUDA?", not "does this model run on CUDA?". Both lead to the same result, but _torch_device is semantically a better fit for a hardware capability check. The only difference is that _get_execution_device() requires config and submodel_type as parameters, while _torch_device does not.

File: invokeai/backend/model_manager/load/model_loaders/flux.py (hunk 2 above)
Comment (Pfannkuchensack) [reply invokeai/backend/model_manager/load/model_loaders/flux.py:142]: GGUF and BnB models are intentionally excluded — they already use their own quantization (typically Q4/Q8), so applying FP8 layerwise casting on top would be redundant and likely conflict with their dequantization logic during inference.

**Comment by Pfannkuchensack** [comment]
[2026-05-09 23:23:46,331]::[ModelLoadService]::INFO --> FP8 layerwise casting enabled for CogView4-6B (storage=float8_e4m3fn, compute=torch.bfloat16, param_size=6768MB)
[2026-05-09 23:23:48,314]::[ModelManagerService]::INFO --> [MODEL CACHE] Loaded model '38d890c8-e623-46f2-8299-6f1a6f97dfee:transformer' (CogView4Transformer2DModel) onto cuda device in 1.90s. Total model size: 6767.89MB, VRAM: 6767.89MB (100.0%)

[2026-05-09 23:25:10,617]::[ModelManagerService]::INFO --> [MODEL CACHE] Loaded model '38d890c8-e623-46f2-8299-6f1a6f97dfee:transformer' (CogView4Transformer2DModel) onto cuda device in 4.11s. Total model size: 12148.13MB, VRAM: 12148.13MB (100.0%)


[2026-05-09 23:28:06,929]::[ModelLoadService]::INFO --> FP8 layerwise casting enabled for sdxl-base-1.0 (storage=float8_e4m3fn, compute=torch.float16, param_size=2449MB)
[2026-05-09 23:28:07,487]::[ModelManagerService]::INFO --> [MODEL CACHE] Loaded model 'ddab47ce-8e0a-49e6-8d67-3d1d48b2848c:unet' (UNet2DConditionModel) onto cuda device in 0.49s. Total model size: 2449.09MB, VRAM: 2449.09MB (100.0%)

[2026-05-09 23:28:53,733]::[ModelManagerService]::INFO --> [MODEL CACHE] Loaded model 'ddab47ce-8e0a-49e6-8d67-3d1d48b2848c:unet' (UNet2DConditionModel) onto cuda device in 0.83s. Total model size: 4897.05MB, VRAM: 4897.05MB (100.0%)


[2026-05-09 23:29:54,400]::[ModelLoadService]::INFO --> FP8 layerwise casting enabled for dreamshaper-8 (storage=float8_e4m3fn, compute=torch.float16, param_size=820MB)
[2026-05-09 23:29:54,604]::[ModelManagerService]::INFO --> [MODEL CACHE] Loaded model '06e7f29b-74ff-4f66-87ce-9b7158e55636:unet' (UNet2DConditionModel) onto cuda device in 0.19s. Total model size: 819.89MB, VRAM: 819.89MB (100.0%)

[2026-05-09 23:29:28,779]::[ModelManagerService]::INFO --> [MODEL CACHE] Loaded model '06e7f29b-74ff-4f66-87ce-9b7158e55636:unet' (UNet2DConditionModel) onto cuda device in 1.69s. Total model size: 1639.41MB, VRAM: 1639.41MB (100.0%)


[2026-05-09 23:33:02,516]::[ModelLoadService]::INFO --> FP8 layerwise casting enabled for controlnet-canny-sdxl-1.0 (storage=float8_e4m3fn, compute=torch.float16, param_size=1193MB)
[2026-05-09 23:33:02,782]::[ModelManagerService]::INFO --> [MODEL CACHE] Loaded model 'd9f95803-6f69-4d43-abc8-c70e1b4a2c2d' (ControlNetModel) onto cuda device in 0.24s. Total model size: 1193.33MB, VRAM: 1193.33MB (100.0%)

[2026-05-09 23:33:42,713]::[ModelManagerService]::INFO --> [MODEL CACHE] Loaded model 'd9f95803-6f69-4d43-abc8-c70e1b4a2c2d' (ControlNetModel) onto cuda device in 0.85s. Total model size: 2386.12MB, VRAM: 2386.12MB (100.0%)

**Comment by Pfannkuchensack** [comment]
[2026-05-09 23:39:48,573]::[ModelLoadService]::INFO --> FP8 layerwise casting enabled for FLUX.2-klein-9B (storage=float8_e4m3fn, compute=torch.bfloat16, param_size=8691MB)
[2026-05-09 23:39:49,800]::[ModelManagerService]::INFO --> [MODEL CACHE] Loaded model '8dbb1dfe-7a08-4eae-8846-2f482e9320f8:transformer' (Flux2Transformer2DModel) onto cuda device in 1.22s. Total model size: 8690.52MB, VRAM: 8690.52MB (100.0%)

[2026-05-09 23:40:20,500]::[ModelManagerService]::INFO --> [MODEL CACHE] Loaded model '8dbb1dfe-7a08-4eae-8846-2f482e9320f8:transformer' (Flux2Transformer2DModel) onto cuda device in 4.34s. Total model size: 17316.02MB, VRAM: 17316.02MB (100.0%)

**Comment by Pfannkuchensack** [comment]
[2026-05-10 00:20:23,041]::[ModelLoadService]::INFO --> FP8 layerwise casting enabled for FLUX.1 dev (storage=float8_e4m3fn, compute=torch.bfloat16, param_size=11350MB)
[2026-05-10 00:20:24,904]::[ModelManagerService]::INFO --> [MODEL CACHE] Loaded model 'fa37dbd1-9681-4d00-a5b2-dc5699237d7f:transformer' (Flux) onto cuda device in 1.84s. Total model size: 11350.07MB, VRAM: 11350.07MB (100.0%)

[2026-05-10 00:21:06,488]::[ModelManagerService]::INFO --> [MODEL CACHE] Loaded model 'fa37dbd1-9681-4d00-a5b2-dc5699237d7f:transformer' (Flux) onto cuda device in 6.37s. Total model size: 22700.13MB, VRAM: 17660.13MB (77.8%)

**Comment by JPPhoto** [comment]
- `invokeai/backend/model_manager/load/load_default.py:98` and `invokeai/app/api/routers/model_manager.py:440` - changing `default_settings.fp8_storage` does not invalidate already-cached models. The cache key is only `model key + submodel_type`, so toggling FP8 on or off in Model Manager silently has no effect until cache clear, eviction, or restart.
- `invokeai/frontend/web/src/features/modelManagerV2/subpanels/ModelPanel/ModelView.tsx:141` and `invokeai/backend/model_manager/load/model_loaders/lora.py:72` - ControlLoRA gets the FP8 Storage toggle, but ControlLoRA loads through `LoRALoader`, which never calls `_apply_fp8_layerwise_casting()`. The setting is persisted on the config but appears to be a no-op for ControlLoRA.
- `invokeai/backend/model_manager/load/load_default.py:233` - the plain `nn.Module` FP8 fallback is not exception-safe. It casts params to compute dtype in a pre-hook and casts back in a normal forward hook; if forward raises, params can remain fp16/bf16, defeating FP8 storage and making cache size accounting stale.
- I didn't find dedicated tests covering FP8 setting persistence, cache reload behavior, ControlLoRA behavior, or exception cleanup.

**Comment by JPPhoto** [comment]
Progress! Here are some more issues or things that need tests to validate current behavior:

- `invokeai/backend/model_manager/load/load_default.py:228` and `invokeai/backend/model_manager/load/model_cache/model_cache.py:337` - FP8 checkpoint models can bypass InvokeAI custom layer behavior, including LoRA patch handling. The FP8 fallback assigns an instance-level `forward` wrapper before `ModelCache.put()` later replaces `torch.nn.Linear` with `CustomLinear`. That replacement shares the old module `__dict__`, preserving the instance `forward`, so `CustomLinear.forward()` is skipped. Scenario: enable FP8 Storage for a FLUX checkpoint and use a LoRA or ControlLoRA patch; the patch path may not run. A minimal reproduction confirmed the wrapped `forward` survives the class swap.
- `invokeai/backend/model_manager/load/model_cache/model_cache.py:881` and `invokeai/app/api/routers/model_manager.py:445` - cache invalidation still misses models locked during the settings update. `drop_model()` skips locked entries, but it does not mark them stale for removal after unlock. Scenario: a long generation is using a model while an admin toggles `fp8_storage` or `cpu_only`; the setting is saved, the locked stale entry survives, and later generations continue reusing it.
- `invokeai/backend/model_manager/load/load_default.py:220` - the plain `nn.Module` FP8 fallback casts every module parameter to FP8, with no skip list for precision-sensitive layers. Diffusers `enable_layerwise_casting()` intentionally skips patterns such as normalization/embedding modules by default, but checkpoint FLUX through the fallback casts all direct params, including FLUX `RMSNorm.scale` in `invokeai/backend/flux/modules/layers.py:63`. Scenario: FP8 Storage on checkpoint FLUX may degrade quality more than the diffusers path.
- `invokeai/backend/model_manager/load/model_cache/model_cache.py:887` - `drop_model()` deletes cache entries without updating `stats.cleared` or firing `_on_cache_models_cleared_callbacks`, unlike `_make_room_internal()`. This is not likely to break inference, but cache stats and any cache-clear observers can miss invalidations caused by model setting changes.

**Comment by JPPhoto** [comment]
Sorry for missing this one...
- `invokeai/backend/model_manager/load/load_default.py:282` dispatches the FP8 wrapper through the original module object after cache custom-layer wrapping, so FP8 plain `nn.Module` models can bypass InvokeAI's `CustomLinear.forward()` path for LoRA patches and device autocast. The wrapper closes over `module` and calls `type(module).forward(module, ...)` at `invokeai/backend/model_manager/load/load_default.py:286`, but production `apply_custom_layers_to_model()` creates a new wrapper object with a shared `__dict__` and installs it on the parent at `invokeai/backend/model_manager/load/model_cache/torch_module_autocast/torch_module_autocast.py:73` and `invokeai/backend/model_manager/load/model_cache/torch_module_autocast/torch_module_autocast.py:99`. The instance `forward` stored in that shared `__dict__` still closes over the old `Linear`, so it calls `torch.nn.Linear.forward`, not `CustomLinear.forward`, and skips the LoRA/patch-aware branch at `invokeai/backend/model_manager/load/model_cache/torch_module_autocast/custom_modules/custom_linear.py:104`. Trigger: enable `fp8_storage` on a plain `nn.Module` loader path such as checkpoint FLUX, cache it via `ModelCache.put()`, then apply LoRA/control patches. I reproduced this in the fresh worktree: after wrapping, the installed child type was `CustomLinear`, but the call path recorded `Linear` on the original object. To expose this issue, add a test that uses real `apply_custom_layers_to_model()` instead of a simulated `__class__` swap, attaches a marker/patch to the resulting `CustomLinear`, and asserts the FP8-wrapped forward reaches `CustomLinear.forward()`.

**Review by JPPhoto — APPROVED** [review]
Looks good, add documentation and then it can go in the next RC so the community can try it out.

**Review by lstein — APPROVED** [review]
Looking forward to trying this.
